packages/synapse-react-client/src/components/doi/CreateOrUpdateDoiModal.tsx

```diff
@@ -0,0 +1,324 @@
+  } else {
+    return DoiResourceTypeGeneral.Dataset
+  }
+}
+
+function convertDoiToFormData(doi: V2Doi): DoiFormData {
+  return {
+    titles: (doi.titles || []).map(title => title.title!),
+    creators: (doi.creators || []).map(creator => creator.creatorName!),
+    resourceTypeGeneral: doi.resourceType?.resourceTypeGeneral,
+    publicationYear: doi.publicationYear,
+  }
+}
+
+/**
+ * Returns true if the DOI has fields that are not exposed in the UI, so we can let the user know that any changes
+ * may be destructive to hidden fields.
+ */
+function getWasModifiedViaAPI(doi: V2Doi) {
+  if (doi.creators?.some(creator => !isEmpty(creator.nameIdentifiers))) {
+    return true
+  }
+  return false
+}
+
+function convertFormDataToDoi(formData: DoiFormData): V2Doi {
+  return {
+    titles: formData.titles?.map(title => ({ title })) || [],
+    creators:
+      formData.creators?.map(creator => ({ creatorName: creator })) || [],
+    resourceType: {
+      resourceTypeGeneral: formData.resourceTypeGeneral,
+    },
+    publicationYear: formData.publicationYear,
+  }
+}
+
+export function CreateOrUpdateDoiModal(props: CreateOrUpdateDoiModalProps) {
+  const { open, onClose, objectType, objectId, defaultVersionNumber } = props
+  const [selectedVersionNumber, setSelectedVersionNumber] =
+    useState(defaultVersionNumber)
+
+  const [formData, setFormData] = useState<DoiFormData>({
+    titles: [''],
+    creators: [''],
+  })
+  const { data: currentUser, isLoading: isLoadingCurrentUser } =
+    useGetCurrentUserProfile()
+  const { data: entity, isLoading: isLoadingEntity } = useGetEntity(
+    objectId,
+    defaultVersionNumber,
+    {
+      enabled: objectType === 'ENTITY',
+      throwOnError: true,
+    },
+  )
+  const { data: doi, isLoading: isLoadingDoi } = useGetDOI(
+    objectId,
+    selectedVersionNumber,
+    objectType,
+    {
+      staleTime: Infinity,
+      throwOnError: true,
+    },
+  )
+
+  const doiCanBeAppliedToVersion =
+    objectType === 'ENTITY' && entity && isVersionableEntity(entity)
+
+  const { data: entityVersions } = useGetVersions(
+    objectId,
+    undefined,
+    undefined,
+    {
+      enabled: doiCanBeAppliedToVersion,
+    },
+  )
+
+  const entityType = entity
+    ? convertToEntityType(entity.concreteType)
+    : EntityType.FILE
+  const entityTypeDisplay = entityTypeToFriendlyName(entityType)
+  let versionHelpMarkdown = `A DOI can be associated with a specific version of this ${entityTypeDisplay}.
+  
+  Versioned DOIs will link to the specified version of the ${entityTypeDisplay}.
+  
+  A DOI without a version will always link to the newest version of this ${entityTypeDisplay},
+     so the data that someone retrieves using the DOI may change over time.`
+  if (entity && isTable(entity)) {
+    versionHelpMarkdown += `\n\nTo create a DOI that will always link to the current set of data in the ${entityTypeDisplay},
+     create a new version and mint a DOI for that version.`
```

Comment:
If it's a Table, should this say "create a new snapshot" instead of "create a new version"?  Version works for me, but I know we have other UI that refers to creating Snapshots

Reply:
This is the same text we have in the GWT implementation. We have some utilities to handle the different cases (e.g. Tables have "snapshots" but Datasets have "versions"). I will see if it's easy to integrate those utilities